File: swift/Sources/ClickMan/Queue.swift
import ClickManCore
import Foundation
import os

/// The device queue of the Rust core, behind `clickman.h`. The core serializes
/// calls itself, so the handle may be used from any thread.
final class Queue: @unchecked Sendable {
    private let handle: OpaquePointer
    private let logger: Logger

    init(storage: URL, configuration: [String: Int], logger: Logger) throws {
        try FileManager.default.createDirectory(
            at: storage.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        let config = String(decoding: try JSONSerialization.data(withJSONObject: configuration), as: UTF8.self)

        guard let handle = clickman_open(storage.path, config) else {
            throw ClickManError.storage(Self.lastError)
        }
        self.handle = handle
        self.logger = logger
    }

    deinit {
        clickman_close(handle)
    }

    func identify(_ externalId: String) {
        check(clickman_identify(handle, externalId), "identify")
    }

    func reset() {
        check(clickman_reset(handle), "reset")
    }

    func setContext(_ json: String) {
        check(clickman_set_context(handle, json), "set the context")
    }

    func setTraits(_ json: String) {
        check(clickman_set_traits(handle, json), "set the traits")
    }

    func appLaunched(version: String, build: String, at date: Date) {
        check(clickman_app_launched(handle, version, build, Self.milliseconds(date)), "record the launch")
    }

    func track(_ event: String, properties json: String, at timestamp: Date) {
        check(clickman_track(handle, event, json, Self.milliseconds(timestamp)), "track \(event)")
    }

    func takeBatch(now: Date, force: Bool) -> Batch? {
        var batch = clickman_batch()
        let taken = clickman_take_batch(handle, Self.milliseconds(now), force, &batch)
        guard taken == 1 else {
            check(taken, "take a batch")
            return nil
        }
        defer { clickman_buf_free(batch.body) }

        return Batch(id: batch.id, events: Int(batch.events), body: Data(bytes: batch.body.ptr, count: batch.body.len))
    }

    func complete(_ batch: Batch, status: Int, retryAfter: TimeInterval?, now: Date) {
        let retryAfter = retryAfter.map { Int64($0 * 1000) } ?? -1
        check(
            clickman_complete_batch(handle, batch.id, Int32(clamping: status), retryAfter, Self.milliseconds(now)),
            "complete batch \(batch.id)"
        )
    }

    var pending: Int {
        Int(clickman_pending(handle))
    }

    private func check(_ result: Int32, _ action: String) {
        if result < 0 {
            logger.error("ClickMan could not \(action, privacy: .public): \(Self.lastError, privacy: .public)")
        }
    }

    private static var lastError: String {
        String(cString: clickman_last_error())
    }

    private static func milliseconds(_ date: Date) -> Int64 {
        Int64((date.timeIntervalSince1970 * 1000).rounded())
    }
}

struct Batch: Sendable {
    let id: UInt64
    let events: Int
    let body: Data
}

File: kotlin/clickman/src/main/kotlin/com/lockvoid/clickman/Queue.kt
package com.lockvoid.clickman

import com.sun.jna.Pointer
import java.io.Closeable
import java.io.File
import java.util.concurrent.locks.ReentrantReadWriteLock
import java.util.logging.Logger
import kotlin.concurrent.read
import kotlin.concurrent.write

internal class Batch(val id: Long, val events: Int, val body: ByteArray)

/**
 * The device queue of the Rust core. The core serializes calls itself, so any
 * thread may call; the lock only keeps a call from reaching the core once
 * [close] has freed it.
 */
internal class Queue(storage: File, configurationJson: String, private val logger: Logger) : Closeable {
    private val native = ClickManNative.INSTANCE
    private val handle: Pointer
    private val lock = ReentrantReadWriteLock()
    private var closed = false

    init {
        storage.absoluteFile.parentFile.mkdirs()
        handle = native.clickman_open(storage.path, configurationJson)
            ?: throw ClickManException("ClickMan could not open ${storage.path}: ${native.clickman_last_error()}")
    }

    fun identify(externalId: String) = call("identify") { native.clickman_identify(handle, externalId) }

    fun reset() = call("reset") { native.clickman_reset(handle) }

    fun setContext(json: String) = call("set the context") { native.clickman_set_context(handle, json) }

    fun setTraits(json: String) = call("set the traits") { native.clickman_set_traits(handle, json) }

    fun appLaunched(version: String, build: String, nowMs: Long) =
        call("record the launch") { native.clickman_app_launched(handle, version, build, nowMs) }

    fun track(event: String, propertiesJson: String, timestampMs: Long) =
        call("track $event") { native.clickman_track(handle, event, propertiesJson, timestampMs) }

    fun takeBatch(nowMs: Long, force: Boolean): Batch? =
        lock.read {
            if (closed) {
                return null
            }

            val out = ClickManBatch()
            val taken = native.clickman_take_batch(handle, nowMs, force, out)
            if (taken != 1) {
                report(taken, "take a batch")
                return null
            }

            val pointer = checkNotNull(out.body.ptr) { "a taken batch has a body" }
            val body = pointer.getByteArray(0, out.body.len.toInt())
            native.clickman_buf_free(ClickManBuf.ByValue().also {
                it.ptr = out.body.ptr
                it.len = out.body.len
            })
            Batch(out.id, out.events, body)
        }

    fun complete(batch: Batch, status: Int, retryAfterMs: Long, nowMs: Long) =
        call("complete batch ${batch.id}") { native.clickman_complete_batch(handle, batch.id, status, retryAfterMs, nowMs) }

    val pending: Int
        get() = lock.read { if (closed) 0 else native.clickman_pending(handle).toInt() }

    override fun close() {
        lock.write {
            if (!closed) {
                closed = true
                native.clickman_close(handle)
            }
        }
    }

    private fun call(action: String, work: () -> Int) {
        lock.read {
            if (closed) {
                logger.warning("ClickMan could not $action: it is closed")
                return
            }
            report(work(), action)
        }
    }

    private fun report(result: Int, action: String) {
        if (result < 0) {
            logger.warning("ClickMan could not $action: ${native.clickman_last_error()}")
        }
    }
}
